test/unit/BCoWHelper.t.sol

```diff
@@ -0,0 +1,156 @@
+    vm.mockCall(tokens[1], abi.encodePacked(IERC20.balanceOf.selector), abi.encode(priceVector[1]));
+
+    factory.mock_call_APP_DATA(bytes32('appData'));
+    helper = new BCoWHelper(address(factory));
+  }
+
+  function test_ConstructorWhenCalled(bytes32 _appData) external {
+    factory.expectCall_APP_DATA();
+    factory.mock_call_APP_DATA(_appData);
+    helper = new BCoWHelper(address(factory));
+    // it should set factory
+    assertEq(helper.factory(), address(factory));
+    // it should set app data from factory
+    assertEq(helper.APP_DATA(), _appData);
+  }
+
+  function test_TokensRevertWhen_PoolIsNotRegisteredInFactory() external {
+    // it should revert
+    vm.expectRevert(ICOWAMMPoolHelper.PoolDoesNotExist.selector);
+    helper.tokens(invalidPool);
+  }
+
+  function test_TokensRevertWhen_PoolHasLessThan2Tokens() external {
+    address[] memory invalidTokens = new address[](1);
+    invalidTokens[0] = makeAddr('token0');
+    pool.set__tokens(invalidTokens);
+    // it should revert
+    vm.expectRevert(ICOWAMMPoolHelper.PoolDoesNotExist.selector);
+    helper.tokens(address(pool));
+  }
+
+  function test_TokensRevertWhen_PoolHasMoreThan2Tokens() external {
+    address[] memory invalidTokens = new address[](3);
+    invalidTokens[0] = makeAddr('token0');
+    invalidTokens[1] = makeAddr('token1');
+    invalidTokens[2] = makeAddr('token2');
+    pool.set__tokens(invalidTokens);
+    // it should revert
+    vm.expectRevert(ICOWAMMPoolHelper.PoolDoesNotExist.selector);
+    helper.tokens(address(pool));
+  }
+
+  function test_TokensRevertWhen_PoolTokensHaveDifferentWeights() external {
+    pool.mock_call_getNormalizedWeight(tokens[0], VALID_WEIGHT);
+    pool.mock_call_getNormalizedWeight(tokens[1], VALID_WEIGHT + 1);
+
+    vm.expectRevert(ICOWAMMPoolHelper.PoolDoesNotExist.selector);
+    // it should revert
+    helper.tokens(address(pool));
+  }
+
+  function test_TokensWhenPoolIsSupported() external view {
+    // it should return pool tokens
+    address[] memory returned = helper.tokens(address(pool));
+    assertEq(returned[0], tokens[0]);
+    assertEq(returned[1], tokens[1]);
+  }
+
+  function test_OrderRevertWhen_ThePoolIsNotSupported() external {
+    // it should revert
+    vm.expectRevert(ICOWAMMPoolHelper.PoolDoesNotExist.selector);
+    helper.order(invalidPool, priceVector);
+  }
+
+  function test_OrderWhenThePoolIsSupported(bytes32 domainSeparator) external {
+    // it should query the domain separator from the pool
+    pool.expectCall_SOLUTION_SETTLER_DOMAIN_SEPARATOR();
+    pool.mock_call_SOLUTION_SETTLER_DOMAIN_SEPARATOR(domainSeparator);
+
+    (
+      GPv2Order.Data memory order_,
+      GPv2Interaction.Data[] memory preInteractions,
+      GPv2Interaction.Data[] memory postInteractions,
+      bytes memory sig
+    ) = helper.order(address(pool), priceVector);
+
+    // it should return a valid pool order
+    assertEq(order_.receiver, GPv2Order.RECEIVER_SAME_AS_OWNER);
```

Comment:
no asserts on the amounts, they are recovered from the library which cannot be mocked, we're just testing here that what the library returns is, as for our standards, a valid order (also checked on integration test)

Reply:
I think it's worth it to precompute some values to make sure we're not passing parameters to the library in the inverse order (or other silly-but-consequential bugs like that). This won't block merging since we're already doing it for the integrations, but is still desireable. (related to https://github.com/defi-wonderland/balancer-v1-amm/pull/121/files#r1678466826)

Reply:
this is indeed tough because we don't have a precomputed amount that we're trading against the pool, but the helper is the one that says how much, we do have an expectation and is that at the end, the spot price should be similar to the pricing vector, will try that one